alpha/veneer/basic/basic.go

```diff
@@ -0,0 +1,313 @@
+package basic
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"github.com/operator-framework/operator-registry/pkg/image/containerdregistry"
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+)
+
+var ErrNotAllowed = errors.New("not allowed")
+
+type BasicVeneer struct {
+	Registry image.Registry
+
+	GenerateSemverChannels bool
+}
+
+func nullLogger() *logrus.Entry {
+	logger := logrus.New()
+	logger.SetOutput(ioutil.Discard)
+	return logrus.NewEntry(logger)
+}
+
+func (v BasicVeneer) Ingest(ctx context.Context, files []string) (*declcfg.DeclarativeConfig, error) {
+	if v.Registry == nil {
+		reg, err := v.createRegistry()
+		if err != nil {
+			return nil, fmt.Errorf("create registry: %v", err)
+		}
+		defer reg.Destroy()
+		v.Registry = reg
+	}
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, f := range files {
+		cfg, err := v.renderReference(ctx, f)
+		if err != nil {
+			return nil, fmt.Errorf("render reference %q: %w", f, err)
+		}
+		renderBundleObjects(cfg)
+
+		for _, b := range cfg.Bundles {
+			sort.Slice(b.RelatedImages, func(i, j int) bool {
+				return b.RelatedImages[i].Image < b.RelatedImages[j].Image
+			})
+		}
+
+		cfgs = append(cfgs, *cfg)
+	}
+
+	return combineConfigs(cfgs), nil
+}
+
+func (v BasicVeneer) createRegistry() (*containerdregistry.Registry, error) {
+	cacheDir, err := os.MkdirTemp("", "veneer-registry-")
+	if err != nil {
+		return nil, fmt.Errorf("create tempdir: %v", err)
+	}
+
+	reg, err := containerdregistry.NewRegistry(
+		containerdregistry.WithCacheDir(cacheDir),
+
+		// The containerd registry impl is somewhat verbose, even on the happy path,
+		// so discard all logger logs. Any important failures will be returned from
+		// registry methods and eventually logged as fatal errors.
+		containerdregistry.WithLog(nullLogger()),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return reg, nil
+}
+
+func (v BasicVeneer) renderReference(ctx context.Context, ref string) (*declcfg.DeclarativeConfig, error) {
+	if stat, serr := os.Stat(ref); serr == nil {
+		if stat.IsDir() {
+			return nil, fmt.Errorf("cannot ingest veneer directory: %w", ErrNotAllowed) // BUGBUG enhance to walk and eat all files in turn, heeding ignore file
```

Comment:
I'd advocate for a more descriptive error message and variable name here. Something like

```suggestion
			return nil, ErrCannotIngestDirectory // BUGBUG enhance to walk and eat all files in turn, heeding ignore file
```

Where `ErrCannotIngestDirectory` is the exact same error you defined above (maybe with a "why", but that's up to you) ^

```go
var ErrCannotIngestDirectory = errors.New("cannot ingest veneer directory") // Also potentially add a why here or in the return.
```

Reply:
IMO, the motivating factor for specific error variables like that is "do I need to check for that specific error from a calling scope?" 

If the answer is yes, it's better to create the variable such that callers can use `errors.Is(err, ErrCannotIngestDirectory)` instead of `strings.Contain(err.Error(), "cannot ingest directory")

If the answer is no, it's better to just use `fmt.Errorf`. Also, it should be noted that using `%w` in the error message format instead of `%v` is a defacto public API because it enables callers to use `errors.Is` and `errors.As` to match against wrapped errors. I generally always default to `%v` for this reason. It's simple to change a `%v` to a `%w` when necessary without breaking any callers, but not the other way around. See https://go.dev/blog/go1.13-errors#whether-to-wrap

Reply:
The error would be needed in one place, and on an alpha-level tool I'm disinclined to invest a lot of effort in it.  Based on other comments to unexport it if not needed externally, IMHO this is better off simpler.  
If not, drop a comment and we'll update it.